### PR TITLE
Introduce channel reserve limit

### PR DESCRIPTION
**Pull request by cjamthagen**
In this PR I introduce the channel reserve limit. Currently only the remote peer's reserve limit is enforced when we receive an HTLC from them. 

Checking that the local reserve limit isn't breached didn't seem to look good in lnwallet/channel.go, but rather after having reviewed #118 (and assuming this will be merged), having Bandwidth() of the htlcmanager return the expected local balance, with the reserve limit subtracted from it, would be a better solution. Feedback on this would be much appreciated.

### CODE REVIEW

File: lnwallet/channel.go [hunk 1]
```diff
@@ -546,6 +547,13 @@ type LightningChannel struct {
 	// way to lookup the original PaymentDescriptor.
 	rHashMap map[PaymentHash][]*PaymentDescriptor
 
+	// [Our|Their]ExpectedBalance reflects the available balance both locally
+	// and remotely. Every time an update happens such as adding or receiving
+	// a HTLC, these values are updated immediately to reflect the available
+	// balance and make sure they don't go below the channel reserve limit.
+	OurExpectedBalance   btcutil.Amount
```
Comment (Roasbeef) [inline lnwallet/channel.go]: As pointed out in my PR-level comment, I think this is the wrong level of abstraction to enforce these limits.

Comment (Roasbeef) [inline lnwallet/channel.go]: Also as a style comment: these variables don't need to be publicly exported.

Comment (Roasbeef) [inline lnwallet/channel.go]: Also with the semantics of the new state machine, updates in the log aren't guaranteed to be included in the next state transition if both sides add to the log concurrently

**Comment by cjamthagen** [comment]
> Hmm, I think this functionality can better be implemented by encapsulating much of the logic within the htlcSwitch itself.

Hmm.. I did consider this when I first started. Can't remember why I went another way..

> This would involve adding a new bandwidthReserve attribute to the (as it's currently called) link struct.

I implemented this by adjusting the `availableBandwidth` directly when the link is created in order to avoid adding this new attribute. https://github.com/lightningnetwork/lnd/pull/122/commits/5ede96d52958928416831c97cdd738008d040747

> For the remote reserver, I'm not so sure we need to actively employ any new policy on our end.

Since the point of the reserve limit is to deter our peer from closing the channel at an invalid state where they have the most to gain and very little (or nothing) to lose, it seems like its in our best interest to actively check that they don't go below their reserve limit instead of trusting them not to do it.

> However, if we want to add a preemptive measure, when processing a new HLTC add, if the add puts them below their set reserve, then the HTLC should be rejected.

Does my changes that implement this in lnwallet/channel.go make sense, or should they also go elsewhere?

File: htlcswitch.go [hunk 2]
```diff
@@ -528,7 +528,7 @@ func (h *htlcSwitch) handleRegisterLink(req *registerLinkMsg) {
 	chanPoint := req.linkInfo.ChannelPoint
 	newLink := &link{
 		capacity:           req.linkInfo.Capacity,
-		availableBandwidth: int64(req.linkInfo.LocalBalance),
+		availableBandwidth: int64(req.linkInfo.LocalBalance - req.linkInfo.ChannelReserve),
```
Comment (Roasbeef) [inline htlcswitch.go]: Very minimal change, I dig it!

File: lnwallet/channel.go [hunk 3]
```diff
@@ -547,6 +548,12 @@ type LightningChannel struct {
 	// way to lookup the original PaymentDescriptor.
 	rHashMap map[PaymentHash][]*PaymentDescriptor
 
+	// theirExpectedBalance reflects the available balance both locally
```
Comment (Roasbeef) [inline lnwallet/channel.go]: I don't think this extra variable is required in order to enforce channel reserve semantics within the state machine itself. Instead, `validateCommitmentSanity` can be extended slightly to calculate the "expected balance" in real-time. This consolidates the commitment sanity checks (too many htlc's, reserve violation, etc) in once place, and will also still maintain its correctness once we move back to the sliding-window based commitment protocol.

File: lnwallet/channel_test.go [hunk 4]
```diff
@@ -842,9 +851,217 @@ func TestCheckHTLCNumberConstraint(t *testing.T) {
...
+	// Bob creates an HTLC which leaves him with the minimum channel
+	// reserve allowed.
+	htlc, preimage = createHTLC(0, bobAmount-bobChannelReserve)
+	if _, err := bobChannel.AddHTLC(htlc); err != nil {
+		t.Fatalf("bob unable to add htlc: %v", err)
+	}
+	if _, err := aliceChannel.ReceiveHTLC(htlc); err != nil {
+		t.Fatalf("alice unable to receive htlc: %v", err)
+	}
+	if err := forceStateTransition(aliceChannel, bobChannel); err != nil {
```
Comment (Roasbeef) [inline lnwallet/channel_test.go]: Due to the new state machine, the order of the arguments here will need to be swapped.

**Comment by Roasbeef** [comment]
Closing for now as this PR is dated and the commitment state machine and `htlcswitch` have changed drastically since it was proposed. Also reserves as specified within the specification must be selected by the _opposite_ party (so Alice selects Bob's reserve).

**Pull request closed without merging**